include/seqan3/range/detail/random_access_iterator.hpp

```diff
@@ -44,214 +44,218 @@
-
-    //!\brief Construct by host, default position pointer with 0.
-    explicit constexpr random_access_iterator(container_type & host) noexcept : host{&host} {}
-
-    //!\brief Construct by host and explicit position.
-    constexpr random_access_iterator(container_type & host, position_type const pos) noexcept : host{&host}, pos{pos} {}
-
+     */
+    //!\brief Default constructor.
+    constexpr random_access_iterator_base() = default;
     //!\brief Copy constructor.
-    constexpr random_access_iterator(random_access_iterator const &) = default;
-
+    constexpr random_access_iterator_base(random_access_iterator_base const &) = default;
     //!\brief Copy construction via assignment.
-    constexpr random_access_iterator & operator=(random_access_iterator const &) = default;
-
+    constexpr random_access_iterator_base & operator=(random_access_iterator_base const &) = default;
     //!\brief Move constructor.
-    constexpr random_access_iterator (random_access_iterator &&) = default;
-
+    constexpr random_access_iterator_base (random_access_iterator_base &&) = default;
     //!\brief Move assignment.
-    constexpr random_access_iterator & operator=(random_access_iterator &&) = default;
-
+    constexpr random_access_iterator_base & operator=(random_access_iterator_base &&) = default;
     //!\brief Use default deconstructor.
-    ~random_access_iterator() = default;
+    ~random_access_iterator_base() = default;
+
+    //!\brief Construct by host, default position pointer with 0.
+    explicit constexpr random_access_iterator_base(range_type & host) noexcept : host{&host} {}
+    //!\brief Construct by host and explicit position.
+    constexpr random_access_iterator_base(range_type & host, position_type const pos) noexcept :
+        host{&host}, pos{pos}
+    {}
 
     //!\brief Constructor for const version from non-const version.
-    template <typename t>
+    template <typename range_type2, typename derived_type2>
     //!\cond
-        requires std::is_same_v<std::remove_const_t<container_type>, t> &&
-                 std::is_same_v<container_type, std::add_const_t<t>>
+        requires std::is_const_v<range_type> && !std::is_const_v<range_type2> &&
+                 std::is_same_v<std::remove_const_t<range_type>, range_type2>
     //!\endcond
-    constexpr random_access_iterator(random_access_iterator<t> const & rhs) noexcept :
+    constexpr random_access_iterator_base(random_access_iterator_base<range_type2, derived_type2> const & rhs) noexcept :
         host{rhs.host}, pos{rhs.pos}
     {}
     //!\}
 
     /*!\name Comparison operators
-     * \brief Compares only the absolute position of two iterators.
+     * \brief seqan3::detail::random_access_iterator_base operators are used unless specialised in derived type.
      * \{
      */
-    constexpr bool operator==(random_access_iterator const & rhs) const noexcept
+    constexpr bool operator==(derived_type const & rhs) const noexcept
```

Comment:
Now, the rhs, must be of the same type as the lhs it's `derived_type`, right?
Which means that if rhs is an iterator over `container_type const` and lhs is an iterator over a `container_type` without const-qualification, this would cause a compiler error, but it is actually added to the standard to be a valid comparison as long as both iterators are reachable from each other (pointing to the same container).
My solution with the `template template` class would have solved this.
Right now such a behaviour is not tested in the tests below. Or at least I couldn't see it. I suggest, to add this test and see if this solution still works. I am happy to be corrected.